### PR TITLE
feat: LT-2110 - Register OGLs

### PR DESCRIPTION
**Pull request by jfaracik**
* Exporters can now register open general licences

**Misc**
* Improved filtering and interface for licences
* More responsive and scales a lot nicer on phones and tablets

### CODE REVIEW

File: apply_for_a_licence/forms/open_general_licences.py [hunk 1]
```diff
@@ -164,10 +169,37 @@ def open_general_licence_forms(request, **kwargs):
                             [],
                         ),
                     ],
-                    # Submit button removed as it doesn't do anything until LT-2110
-                    buttons=[],
+                    buttons=[
+                        conditional(
+                            selected_open_general_licence.get("registration_required"),
+                            Button("Register", "submit", disable_double_click=True),
```
Comment (BenAndrew-Kainos) [inline apply_for_a_licence/forms/open_general_licences.py:175]: Shouldn't we make disable_double_click default to True?

File: apply_for_a_licence/forms/open_general_licences.py [hunk 2]
```diff
@@ -164,10 +169,37 @@ def open_general_licence_forms(request, **kwargs):
...
         ]
     )
+
+
+def open_general_licence_submit_success_page(request, **kwargs):
+    open_general_licence = get_open_general_licence(request, kwargs["pk"])
+    return success_page(
+        request=request,
+        title="Registration complete",
+        secondary_title="You've successfully registered for **"
```
Comment (BenAndrew-Kainos) [inline apply_for_a_licence/forms/open_general_licences.py:190]: Content

File: apply_for_a_licence/forms/open_general_licences.py [hunk 3]
```diff
@@ -164,10 +169,37 @@ def open_general_licence_forms(request, **kwargs):
...
+        + open_general_licence["case_type"]["reference"]["value"]
+        + " ("
+        + open_general_licence["name"]
+        + ")**",
+        description=ApplicationSuccessPage.DESCRIPTION,
+        what_happens_next=[],
+        includes="includes/open-general-licence.html",
+        additional_context={"licence": open_general_licence},
+        links={
+            "View your open general licences": reverse_lazy("licences:licences")
```
Comment (BenAndrew-Kainos) [inline apply_for_a_licence/forms/open_general_licences.py:200]: Content

File: apply_for_a_licence/views.py [hunk 4]
```diff
@@ -93,3 +98,18 @@ class OpenGeneralLicenceQuestions(MultiFormView):
     def init(self, request, **kwargs):
         self.forms = open_general_licence_forms(request, **kwargs)
         self.action = validate_open_general_licences
+
+    def get_success_url(self):
+        post_open_general_licence_cases(self.request, self.get_validated_data())
```
Comment (BenAndrew-Kainos) [inline apply_for_a_licence/views.py:103]: Would this handle errors? I know we've called validate_open_general_licences before this but there are sometimes unique errors on .save()

File: conf/settings.py [hunk 5]
```diff
@@ -67,6 +67,7 @@
     "conf.middleware.LoggingMiddleware",
     "conf.middleware.ProtectAllViewsMiddleware",
     "conf.middleware.UploadFailedMiddleware",
+    "django.middleware.gzip.GZipMiddleware",
```
Comment (BenAndrew-Kainos) [inline conf/settings.py:70]: Nice :)

File: core/services.py [hunk 6]
```diff
@@ -308,3 +315,8 @@ def get_open_general_licences(request, convert_to_options=False, case_type=None,
 
 def get_open_general_licence(request, pk):
     return get(request, OPEN_GENERAL_LICENCES_URL + str(pk)).json()
+
+
+def post_open_general_licence_cases(request, json):
+    data = post(request, "/licences/open-general-licences/", json)
```
Comment (BenAndrew-Kainos) [inline core/services.py:321]: constants

File: licences/helpers.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+from lite_forms.components import Option
+
+
+def get_potential_ogl_control_list_entries(open_general_licences):
+    control_list_entries = []
+
+    print(open_general_licences)
```
Comment (BenAndrew-Kainos) [inline licences/helpers.py]: remove

File: licences/helpers.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    for open_general_licence in open_general_licences:
+        for control_list_entry in open_general_licence["control_list_entries"]:
+            if control_list_entry not in control_list_entries:
```
Comment (BenAndrew-Kainos) [inline licences/helpers.py:9]: You could just make control_list_entries a set and use .add()

File: licences/helpers.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    return return_value
+
+
+def get_potential_ogl_countries(open_general_licences):
+    countries = []
+
+    for open_general_licence in open_general_licences:
+        for country in open_general_licence["countries"]:
+            if country not in countries:
```
Comment (BenAndrew-Kainos) [inline licences/helpers.py:24]: same here as above

File: licences/helpers.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    return return_value
+
+
+def get_potential_ogl_sites(open_general_licences):
+    sites = []
+
+    for open_general_licence in open_general_licences:
+        for registration in open_general_licence["registrations"]:
+            if registration["site"] not in sites:
```
Comment (BenAndrew-Kainos) [inline licences/helpers.py:39]: same here as above

File: licences/services.py [hunk 11]
```diff
@@ -7,12 +7,19 @@ def get_licences(
...
-        data = get(request, LICENCES_URL + NLR_URL + convert_parameters_to_query_params(locals()))
+        response = get(request, LICENCES_URL + NLR_URL + convert_parameters_to_query_params(locals()))
     else:
-        data = get(request, LICENCES_URL + convert_parameters_to_query_params(locals()))
-    return data.json()
+        response = get(request, LICENCES_URL + convert_parameters_to_query_params(locals()))
+    return response.json()
+
+
+def get_nlr_licences(
```
Comment (BenAndrew-Kainos) [inline licences/services.py:16]: No such thing as a no licence required licence. Please rename to get NLR letters

File: licences/services.py [hunk 12]
```diff
@@ -7,12 +7,19 @@ def get_licences(
...
+    request, page=1, reference=None, clc=None, country=None, end_user=None, active_only=None,
+):
+    response = get(request, LICENCES_URL + NLR_URL + convert_parameters_to_query_params(locals()))
```
Comment (BenAndrew-Kainos) [inline licences/services.py:19]: Did you put NLR's on the Licence endpoint?!?

File: licences/views.py [hunk 13]
```diff
@@ -4,62 +4,125 @@
...
+    type = None
+    data = None
+    filters = None
+    template = None
+    page = 1
+
+    def get_licences(self):
+        self.data = get_licences(self.request, **self.request.GET)
+        self.filters = [
+            TextInput(name="reference", title=LicencesList.Filters.REFERENCE,),
```
Comment (BenAndrew-Kainos) [inline licences/views.py:40]: Can we move these filters to a shared function so the code isn't duplicated

File: licences/views.py [hunk 14]
```diff
@@ -4,62 +4,125 @@
...
-                    ),
-                ),
-                HiddenField(name="licence_type", value=licence_type),
-                HiddenField(name="page", value=page_no),
-            ]
-        )
+    def get(self, request, **kwargs):
+        self.type = request.GET.get("licence_type", "licences")
+        self.page = int(request.GET.get("page", 1))
+        getattr(self, f"get_{self.type}", self.get_licences)()  # Set template properties
```
Comment (BenAndrew-Kainos) [inline licences/views.py:109]: Could you elaborate on this line, I'm still really confused what this is doing

File: templates/includes/open-general-licence.html [hunk 15]
```diff
@@ -0,0 +1,26 @@
+<h2 class="govuk-heading-m govuk-!-margin-top-6">Sites</h2>
+<table class="govuk-table">
+	<thead class="govuk-table__head">
+		<tr class="govuk-table__row">
+			<th scope="col" class="govuk-table__header">
+				Reference
```
Comment (BenAndrew-Kainos) [inline templates/includes/open-general-licence.html:6]: content here and below

File: templates/layouts/base.html [hunk 16]
```diff
@@ -17,6 +17,11 @@
 	<link rel="apple-touch-icon" sizes="152x152" href="{% static 'images/govuk-apple-touch-icon-152x152.png' %}">
 	<link rel="apple-touch-icon" href="{% static 'images/govuk-apple-touch-icon.png' %}">
 
+	<link rel="dns-prefetch" href="https://www.google-analytics.com">
+	<link rel="dns-prefetch" href="https://www.googletagmanager.com">
+	<link href="https://www.google-analytics.com" rel="preconnect" crossorigin>
```
Comment (BenAndrew-Kainos) [inline templates/layouts/base.html:22]: Has it been confirmed we are ok to use google-analytics?

File: templates/licences/licences.html [hunk 17]
```diff
@@ -13,37 +13,28 @@ <h1 class="govuk-heading-l">{% block title %}{% lcs "licences.LicencesList.TITLE
...
-	{% if licences.results %}
+	{% if data.results %}
 		<table class="govuk-table">
 			<thead class="govuk-table__head">
 				<tr class="govuk-table__row">
-					<th class="govuk-table__header" scope="col">{% lcs "licences.LicencesList.Table.APPLICATION_COLUMN" %}</th>
-					<th class="govuk-table__header" scope="col">{% lcs "licences.LicencesList.Table.LICENCE_COLUMN" %}</th>
-					<th class="govuk-table__header govuk-!-width-one-third" scope="col">{% lcs "licences.LicencesList.Table.GOODS_COLUMN" %}</th>
-					<th class="govuk-table__header govuk-!-width-one-third" scope="col">{% lcs "licences.LicencesList.Table.DESTINATION_COLUMN" %}</th>
+					<th class="govuk-table__header" scope="col">Reference</th>
```
Comment (BenAndrew-Kainos) [inline templates/licences/licences.html:24]: content

File: templates/licences/licences.html [hunk 18]
```diff
@@ -52,9 +43,9 @@ <h1 class="govuk-heading-l">{% block title %}{% lcs "licences.LicencesList.TITLE
 							{% for good in licence.application.goods|slice:":3" %}
 								<div class="app-expanded-row__item--invert">
 									{% if good.good %}
-										{{ forloop.counter }}. {% include 'includes/control-list-entries.html' with control_list_entries=good.good.control_list_entries %}: {{ good.good.description|abbreviate_string:30 }}
```
Comment (BenAndrew-Kainos) [inline templates/licences/licences.html:55]: How comes we're not abbreviating anymore?

File: templates/licences/open-general-licences.html [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+	<div class="lite-app-bar">
+		<div class="lite-app-bar__content">
+			<h1 class="govuk-heading-l">{% block title %}{% lcs "licences.LicencesList.TITLE" %}{% endblock %}</h1>
+		</div>
+	</div>
+
+	{% include 'includes/tabs.html' %}
+
+	<p class="lite-filters__hint-text lite-filters__hint-text--warning">
+		Read {{ 'https://www.gov.uk/government/publications/compliance-code-of-practice'|linkify:'guidance on compliance and enforcement of export controls' }}
```
Comment (BenAndrew-Kainos) [inline templates/licences/open-general-licences.html:19]: Please change this to a markdown link so we can put it in content

File: templates/licences/open-general-licences.html [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+									<li>
+										{{ licence.description }}
+									</li>
+									<li>
+										{{ licence.url|linkify:'Read more on GOV.UK' }}
+									</li>
+								</ul>
+							</div>
+							<div class="govuk-grid-column-one-third govuk-!-margin-bottom-6">
+								<p class="govuk-heading-s">More information</p>
```
Comment (BenAndrew-Kainos) [inline templates/licences/open-general-licences.html:52]: content

File: templates/licences/open-general-licences.html [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+											{{ licence.control_list_entries|length }} control list entr{{ licence.control_list_entries|pluralize:'y,ies' }}
+										</span>
+									</li>
+									<li>
+										<span data-definition-title="{{ licence.case_type.reference.value }} ({{ licence.name }})" data-definition-subtitle="{{ licence.countries|length }} countr{{ open_general_licence.licence|pluralize:'y,ies' }}" data-definition-list="{% for country in licence.countries %}{{ country.name }}{% if not forloop.last %}, {% endif %}{% endfor %}">
+											{{ licence.countries|length }} countr{{ licence.countries|pluralize:'y,ies' }}
+										</span>
+									</li>
+									<li>
+										Documentation held at:
```
Comment (BenAndrew-Kainos) [inline templates/licences/open-general-licences.html:65]: content

File: templates/licences/open-general-licences.html [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+									</li>
+								</ul>
+							</div>
+						</div>
+						<table class="govuk-table">
+							<caption class="govuk-table__caption govuk-!-margin-bottom-3">Sites</caption>
+							<thead class="govuk-table__head">
+								<tr class="govuk-table__row">
+									<th scope="col" class="govuk-table__header">
+										Reference
```
Comment (BenAndrew-Kainos) [inline templates/licences/open-general-licences.html:80]: content in heaaders

File: licences/services.py (hunk 12 above)
Comment (jfaracik) [reply licences/services.py:19]: That was done in a previous story.

File: templates/layouts/base.html (hunk 16 above)
Comment (jfaracik) [reply templates/layouts/base.html:22]: We've been using Google Analytics for around a year now. https://web.dev/uses-rel-preconnect/

File: licences/views.py [hunk 23]
```diff
@@ -4,62 +4,125 @@
...
+                title=LicencesList.Filters.CLC,
+                options=get_control_list_entries(self.request, convert_to_options=True),
+            ),
+            AutocompleteInput(
+                name="country",
+                title=LicencesList.Filters.DESTINATION_COUNTRY,
+                options=get_countries(self.request, convert_to_options=True),
+            ),
+            TextInput(name="end_user", title=LicencesList.Filters.DESTINATION_NAME,),
+            Checkboxes(
```
Comment (Mark-Spiers-kainos) [inline licences/views.py:77]: NLRs have no concept of Active, they either exist or don't

File: apply_for_a_licence/views.py (hunk 4 above)
Comment (jfaracik) [reply apply_for_a_licence/views.py:103]: It'll fall over like any other endpoint if there's a service error (not much we can do about that). The endpoint itself is very minimalistic though so I doubt it'd be an issue.

File: licences/views.py (hunk 14 above)
Comment (jfaracik) [reply licences/views.py:109]: Essentially we call functions depending on the licence type, eg call `get_open_general_licences` if self.type is `open_general_licences`. Allows for custom contexts and logic without the need for a mental `if` statement

File: licences/helpers.py (hunk 8 above)
Comment (jfaracik) [reply licences/helpers.py:9]: Dicts aren't hashable so this won't work.

File: templates/licences/licences.html (hunk 18 above)
Comment (jfaracik) [reply templates/licences/licences.html:55]: Mistake.

File: licences/views.py (hunk 13 above)
Comment (jfaracik) [reply licences/views.py:45]: Done.

File: licences/helpers.py (hunk 8 above)
Comment (BenAndrew-Kainos) [reply licences/helpers.py:9]: Oh fair. Didn't realise it was a dict

File: licences/services.py (hunk 12 above)
Comment (BenAndrew-Kainos) [reply licences/services.py:21]: :(

File: licences/views.py (hunk 14 above)
Comment (BenAndrew-Kainos) [reply licences/views.py:86]: Could you add more to the comment please

File: licences/views.py [hunk 24]
```diff
@@ -4,62 +4,102 @@
...
+        sites = get_potential_ogl_sites(self.data)
+        self.filters = [
+            TextInput(name="name", title="name"),
+            Select(name="case_type", title="type", options=OpenGeneralExportLicenceTypes.as_options(),),
+            AutocompleteInput(name="control_list_entry", title="control list entry", options=control_list_entries,),
+            AutocompleteInput(name="country", title="country", options=countries),
+            Select(name="site", title="site", options=sites,),
+            Checkboxes(
+                name="active_only",
+                options=[Option(key=True, value="Only show active")],
```
Comment (BenAndrew-Kainos) [inline licences/views.py:77]: content

File: ui_automation_tests/pages/shared.py [hunk 25]
```diff
@@ -2,7 +2,7 @@
 
 
 class Shared(BasePage):
-    HEADING = ".govuk-heading-xl"  # CSS
+    HEADING = ".govuk-heading-l"  # CSS
```
Comment (BenAndrew-Kainos) [inline ui_automation_tests/pages/shared.py:5]: Bit concerned about this. Why are we even using the heading?

File: ui_automation_tests/step_defs/test_ogel.py [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+    pass
+
+
+@when(parsers.parse('I search for an ogel application of type "{good_type}" for "{country}"'))  # noqa
+def create_standard_application(driver, good_type, country, context):  # noqa
+    ExporterHubPage(driver).click_apply_for_a_licence()
+    ApplyForALicencePage(driver).select_licence_type("export_licence")
+    functions.click_submit(driver)
+    enter_type_of_application(driver, "ogel", context)
+    OgelPage(driver).enter_control_list_entry(good_type)
```
Comment (BenAndrew-Kainos) [inline ui_automation_tests/step_defs/test_ogel.py:24]: OgelPage can be defined once

File: ui_automation_tests/step_defs/test_ogel.py [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+def ogel_licence_created(apply_for_ogel):  # noqa
+    pass
+
+
+@when(parsers.parse('I search for an ogel application of type "{good_type}" for "{country}"'))  # noqa
+def create_standard_application(driver, good_type, country, context):  # noqa
+    ExporterHubPage(driver).click_apply_for_a_licence()
+    ApplyForALicencePage(driver).select_licence_type("export_licence")
+    functions.click_submit(driver)
+    enter_type_of_application(driver, "ogel", context)
```
Comment (BenAndrew-Kainos) [inline ui_automation_tests/step_defs/test_ogel.py:23]: This function should be associated with a Page

File: ui_automation_tests/step_defs/test_ogel.py [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+
+
+@when("I select the created OGEL")
+def select_created_ogel(driver, context):
+    OgelPage(driver).select_created_ogel(context.ogel_id)
+
+
+@when("I agree to the ogel declaration")
+def ogel_declaration(driver):
+    DeclarationPage(driver).agree_to_ogel_conditions()
```
Comment (BenAndrew-Kainos) [inline ui_automation_tests/step_defs/test_ogel.py:35]: Can be declared once

File: ui_automation_tests/pages/shared.py (hunk 25 above)
Comment (phlptt) [reply ui_automation_tests/pages/shared.py:5]: we were already using a heading? this is just changing the class name for it

File: ui_automation_tests/step_defs/test_ogel.py (hunk 27 above)
Comment (phlptt) [reply ui_automation_tests/step_defs/test_ogel.py:23]: it is, it's just how its imported. see line three
from conftest import enter_type_of_application

this is how this enter_type_of_application is used elsewhere

Comment (BenAndrew-Kainos) [reply ui_automation_tests/step_defs/test_ogel.py:23]: But why is it just a random function in conftest. Conftest is for steps. This should be in helpers or something

**Pull request merged**